### PR TITLE
Implement Biocam reader for 3Brain system

### PR DESCRIPTION
**Pull request by alejoe91**
- [x] test files
- [x] tests
- [x] info about gain and offset

### CODE REVIEW

File: neo/rawio/biocamrawio.py [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+        # transform to slice if possible
+        if sorted(channel_indexes) == channel_indexes and np.all(np.diff(channel_indexes) == 1):
+            channel_ids = slice(channel_indexes[0], channel_indexes[0] + len(channel_indexes))
+        return data[:, channel_indexes]
+
+
+def openBiocamFile(filename, mea_pitch, verbose=False):
+    """Open a Biocam hdf5 file, read and return the recording info, pick te correct method to access raw data,
+    and return this to the caller."""
+    try:
```
Comment (samuelgarcia) [inline neo/rawio/biocamrawio.py:121]: Put this a module level no ?

File: neo/rawio/biocamrawio.py [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+    # Get the actual number of channels used in the recording
+    file_format = rf['3BData'].attrs.get('Version')
+    if file_format == 100:
+        nRecCh = len(rf['3BData/Raw'][0])
+    elif (file_format == 101) or (file_format == 102):
+        nRecCh = int(1. * rf['3BData/Raw'].shape[0] / nFrames)
+    else:
+        raise Exception('Unknown data file format.')
+
+    if verbose:
```
Comment (samuelgarcia) [inline neo/rawio/biocamrawio.py:149]: I would remove this verbose stuff here

File: neo/rawio/biocamrawio.py [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+        print('#       signal range: ', recVars['MinVolt'][0], '- ', recVars['MaxVolt'][0])
+        print('# channels: ', nRecCh)
+        print('# frames: ', nFrames)
+        print('# sampling rate: ', samplingRate)
+    # get channel locations
+    r = (rf['3BRecInfo/3BMeaStreams/Raw/Chs'][()]['Row'] - 1) * mea_pitch
+    c = (rf['3BRecInfo/3BMeaStreams/Raw/Chs'][()]['Col'] - 1) * mea_pitch
+    rawIndices = np.vstack((r, c)).T
+    # assign channel numbers
+    chIndices = np.array([(x - 1) + (y - 1) * nCols for (y, x) in rawIndices])
```
Comment (samuelgarcia) [inline neo/rawio/biocamrawio.py:160]: For consistency I would remove all camel case variable name.

File: neo/rawio/biocamrawio.py [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+    def _get_analogsignal_chunk(self, block_index, seg_index, i_start, i_stop,
+                                stream_index, channel_indexes):
+        if i_start is None:
+            i_start = 0
+        if i_stop is None:
+            i_stop = self._num_frames
+
+        data = self._read_function(self._rf, i_start, i_stop, self._num_chanels)
+        # transform to slice if possible
+        if sorted(channel_indexes) == channel_indexes and np.all(np.diff(channel_indexes) == 1):
```
Comment (samuelgarcia) [inline neo/rawio/biocamrawio.py:113]: In neo we already have this mechanism in base class here https://github.com/NeuralEnsemble/python-neo/blob/master/neo/rawio/baserawio.py#L514
So please remove this logic in the rawio.

File: neo/rawio/biocamrawio.py [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+            channel_ids = slice(channel_indexes[0], channel_indexes[0] + len(channel_indexes))
```
Comment (samuelgarcia) [inline neo/rawio/biocamrawio.py:114]: What is this channel_ids ?

File: neo/rawio/biocamrawio.py [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+        read_function = readHDF5t_100
+    elif (file_format == 100) & (signalInv == -1):
+        read_function = readHDF5t_100_i
+    if ((file_format == 101) | (file_format == 102)) & (signalInv == 1):
+        read_function = readHDF5t_101
+    elif ((file_format == 101) | (file_format == 102)) & (signalInv == -1):
+        read_function = readHDF5t_101_i
+    else:
+        raise RuntimeError("File format unknown.")
+    return rf, nFrames, samplingRate, nRecCh, chIndices, file_format, signalInv, rawIndices, read_function
```
Comment (samuelgarcia) [inline neo/rawio/biocamrawio.py:175]: I would use a dict as output.
So we can enhance it more easily.

**Comment by samuelgarcia** [comment]
@alejoe91 : any news here ?
Do you think it will be ready soon or do I move this to 0.11 release ?

**Comment by alejoe91** [comment]
> @alejoe91 : any news here ?
> Do you think it will be ready soon or do I move this to 0.11 release ?

Not this week on my side. So you can move it to next release if you are releasing this week!

File: neo/rawio/biocamrawio.py (hunk 6 above)
Comment (alejoe91) [reply neo/rawio/biocamrawio.py]: done

File: neo/rawio/biocamrawio.py (hunk 5 above)
Comment (alejoe91) [reply neo/rawio/biocamrawio.py]: done

File: neo/rawio/biocamrawio.py (hunk 4 above)
Comment (alejoe91) [reply neo/rawio/biocamrawio.py]: removed

File: neo/rawio/biocamrawio.py (hunk 3 above)
Comment (alejoe91) [reply neo/rawio/biocamrawio.py]: done

File: neo/rawio/biocamrawio.py (hunk 2 above)
Comment (alejoe91) [reply neo/rawio/biocamrawio.py]: removed

File: neo/rawio/biocamrawio.py (hunk 1 above)
Comment (alejoe91) [reply neo/rawio/biocamrawio.py:127]: what do you mean?

File: neo/rawio/biocamrawio.py [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+            i_stop = self._num_frames
+
+        data = self._read_function(self._rf, i_start, i_stop, self._num_chanels)
+        # transform to slice if possible
+        if sorted(channel_indexes) == channel_indexes and np.all(np.diff(channel_indexes) == 1):
+            channel_ids = slice(channel_indexes[0], channel_indexes[0] + len(channel_indexes))
+        return data[:, channel_indexes]
+
+
+def openBiocamFile(filename, mea_pitch, verbose=False):
```
[the remark this answers is not in the document]
Comment (alejoe91) [reply neo/rawio/biocamrawio.py]: done

**Comment by alejoe91** [comment]
Tests depend on this GIN PR: https://gin.g-node.org/NeuralEnsemble/ephy_testing_data/pulls/66

**Comment by alejoe91** [comment]
@JuliaSprenger @samuelgarcia this is ready to review. Also added a test file on GIN (PR still need revision)

**Comment by alejoe91** [comment]
@JuliaSprenger this seems to be failing for a SpikeGLX problem, but the biocam test passes

**Comment by samuelgarcia** [comment]
Hi Alessio.
Cool. Thanks.
I will have a look next week.

**Comment by samuelgarcia** [comment]
@alejoe91 : can you rebase on master and so re run tests ?

**Comment by alejoe91** [comment]
Sure

File: neo/rawio/biocamrawio.py (hunk 1 above)
Comment (alejoe91) [reply neo/rawio/biocamrawio.py]: done

**Pull request merged**